Kotlin Platform refactor

yschimke opened the pull request:
Plan

 * [x] Rename Platform.kt
 * [x] Mechanical Refactor Platform
 * [x] Rename *Platform.kt
 * [x] Mechanical Refactor *Platform
 * [x] Cleanup & Sane JVM conventions (minimal for internal classes)

Later?

 * [x] Basic Idiomatic Kotlin code

JakeWharton commented on okhttp/src/main/java/okhttp3/internal/platform/Platform.kt:
```diff
@@ -0,0 +1,305 @@
+  open fun logCloseableLeak(message: String, stackTrace: Any?) {
+    var logMessage = message
+    if (stackTrace == null) {
+      logMessage += " To see where this was allocated, set the OkHttpClient logger level to FINE: " + "Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);"
```
wrap

JakeWharton commented on okhttp/src/main/java/okhttp3/internal/platform/Platform.kt:
```diff
@@ -0,0 +1,305 @@
+
+      val jdkWithJettyBoot = Jdk8WithJettyBootPlatform.buildIfSupported()
+
+      return jdkWithJettyBoot ?: Platform()
```
probably want to revert this to an `if (jdkWithJettyBoot != null)` so that the comment below can go on a `return Platform()`

JakeWharton commented on mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java:
```diff
@@ -488,15 +488,15 @@ public void processConnection() throws Exception {
 
           if (protocolNegotiationEnabled) {
-            Platform.get().configureTlsExtensions(sslSocket, null, protocols);
+            Platform.Companion.get().configureTlsExtensions(sslSocket, null, protocols);
```
Do we want to use `@JvmStatic` to minimize these changes in unrelated files?

yschimke replied:
Yep, good idea.  I'm going through the set of steps in the description, should have marked as WIP.  Thanks for the early review.

yschimke commented on okhttp/src/main/java/okhttp3/internal/platform/Platform.kt:
```diff
@@ -0,0 +1,305 @@
+
+  companion object {
+    private val PLATFORM = findPlatform()
+    val INFO = 4
```
Isn't that only relevant for fields in the object instead of the companion object?

JakeWharton replied:
No it's the opposite. Constants can only be declared as top-level or properties in an `object`. It makes their value available for inlining at compile time (just like Java's `static final` with a literal initializer).

oldergod commented:
We could also migrate file by file if this becomes too big

yschimke commented on okhttp-tests/src/test/java/okhttp3/internal/platform/Jdk9PlatformTest.java:
```diff
@@ -40,7 +41,10 @@ public void findsAlpnMethods() {
-  public void testToStringIsClassname() {
-    assertThat(new Jdk9Platform(null, null).toString()).isEqualTo("Jdk9Platform");
+  public void testToStringIsClassname() throws NoSuchMethodException {
+    assumeTrue(getPlatform().equals("jdk9"));
```
I was getting a strange NPE locally, but seemed valid given I was faking data.  Not sure we should assume this runs on other platforms.

yschimke commented:
Some spaces to cleanup still, but potentially shippable.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/platform/AndroidPlatform.kt:
data class ?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/platform/ConscryptPlatform.kt:
```diff
@@ -0,0 +1,124 @@
+ */
+class ConscryptPlatform private constructor() : Platform() {
+
+  private// defaults to true, but allow for older versions of conscrypt if still compatible
```
nit: comment location

swankjesse commented on okhttp/src/main/java/okhttp3/internal/platform/ConscryptPlatform.kt:
```diff
@@ -0,0 +1,124 @@
+      return if (sp != null) {
+        readFieldOrNull(sp, X509TrustManager::class.java,
+            "x509TrustManager")
+      } else null
```
style nit: braces on else

swankjesse commented on okhttp/src/main/java/okhttp3/internal/platform/ConscryptPlatform.kt:
```diff
@@ -0,0 +1,124 @@
+
+        if (!Conscrypt.isAvailable()) {
+          null
+        } else ConscryptPlatform()
```
nit: braces on else

swankjesse commented on okhttp/src/main/java/okhttp3/internal/platform/Jdk8WithJettyBootPlatform.kt:
```diff
@@ -0,0 +1,147 @@
+          && String::class.java == returnType && callArgs.size == 1 && callArgs[0] is List<*>) {
+        val peerProtocols = callArgs[0] as List<*>
+        // Pick the first known protocol the peer advertises.
+        val size = peerProtocols.size
```
nit: inline size?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/platform/Jdk9Platform.kt:
```diff
@@ -0,0 +1,94 @@
+      // be used. Observed if you didn't specify SSLParameters.setApplicationProtocols
+      return if (protocol == null || protocol == "") {
+        null
+      } else protocol
```
nit: else braces

yschimke commented:
I may have gone a little heavy on 
- "if to when"
- lift return out
- convert to expression body

yschimke commented:
@swankjesse @JakeWharton What's the current advice on japi for internal? Can't we ignore these?

```
Comparing binary compatibility of  against 
***! MODIFIED CLASS: PUBLIC (<- PACKAGE_PROTECTED) FINAL (<- NON_FINAL) okhttp3.internal.platform.AndroidPlatform  (not serializable)
	---! REMOVED METHOD: PUBLIC(-) STATIC(-) okhttp3.internal.platform.Platform buildIfSupported()
***! MODIFIED CLASS: PUBLIC (<- PACKAGE_PROTECTED) STATIC FINAL okhttp3.internal.platform.AndroidPlatform$CloseGuard  (not serializable)
	***! MODIFIED METHOD: PUBLIC (<- PACKAGE_PROTECTED) FINAL (<- NON_FINAL) java.lang.Object createAndOpen(java.lang.String)
	***! MODIFIED METHOD: PUBLIC (<- PACKAGE_PROTECTED) FINAL (<- NON_FINAL) boolean warnIfOpen(java.lang.Object)
***! MODIFIED CLASS: PUBLIC FINAL (<- NON_FINAL) okhttp3.internal.platform.ConscryptPlatform  (not serializable)
***! MODIFIED CLASS: PUBLIC (<- PACKAGE_PROTECTED) FINAL (<- NON_FINAL) okhttp3.internal.platform.Jdk8WithJettyBootPlatform  (not serializable)
***! MODIFIED CLASS: PUBLIC (<- PACKAGE_PROTECTED) FINAL okhttp3.internal.platform.Jdk9Platform  (not serializable)
***! MODIFIED CLASS: PUBLIC okhttp3.internal.platform.Platform  (not serializable)
	---! REMOVED METHOD: PUBLIC(-) STATIC(-) java.util.List alpnProtocolNames(java.util.List)
	***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) okhttp3.internal.tls.CertificateChainCleaner buildCertificateChainCleaner(javax.net.ssl.SSLSocketFactory)
	***! MODIFIED METHOD: PUBLIC FINAL (<- NON_FINAL) java.lang.String getPrefix()
```

swankjesse commented:
Can you suppress everything in `internal` ?
